src/v/cloud_io/remote.h

```diff
@@ -304,7 +321,69 @@ class remote final
 
     const io_resources& resources() const { return *_resources; }
 
+    /// Event filter class.
+    ///
+    /// The filter can be used to subscribe to subset of events.
+    /// For instance, only to segment downloads and uploads, or to
+    /// events from all subsystems except one.
+    ///
+    /// The filter is a RAII object. It works until the object
+    /// exists. If the filter is destroyed before the notification
+    /// will be received the receiver of the event will see broken
+    /// promise error.
+    class event_filter {
+        friend class remote;
+
+    public:
+        event_filter() = default;
+
+        explicit event_filter(
+          std::unordered_set<api_activity_type> ignored_events)
+          : _events_to_ignore(std::move(ignored_events)) {}
+
+        void add_source_to_ignore(const retry_chain_node* source) {
+            _sources_to_ignore.insert(source);
+        }
+
+        void remove_source_to_ignore(const retry_chain_node* source) {
+            _sources_to_ignore.erase(source);
+        }
+
+        void cancel() {
+            if (_promise.has_value()) {
+                _hook.unlink();
+                _promise.reset();
+            }
+        }
+
+    private:
+        absl::node_hash_set<const retry_chain_node*> _sources_to_ignore;
+        std::unordered_set<api_activity_type> _events_to_ignore;
+        std::optional<ss::promise<api_activity_notification>> _promise{
+          std::nullopt};
+        intrusive_list_hook _hook;
+    };
+
+    /// Return future that will become available on next cloud storage
+    /// api operation.
+    ///
+    /// \note The operations which are trigger notifications are segment upload,
+    /// segment download, segment(s) delete, manifest upload, manifest download.
+    /// The notification is generated before the actual use and does not
+    /// affected by errors. The notification is generated even if the operation
+    /// failed. Also, every retry is generating its own notification.
+    ///
+    /// \param filter is a notification filter which allows to narrow the set of
+    ///        possible notifications by source and type.
+    /// \return the future which will be available after the next cloud storage
+    ///         API operation.
+    ss::future<api_activity_notification> subscribe(event_filter& filter);
```

Comment:
I recall us probably discussing this before but don't remember the outcome.

This notification mechanism makes it impossible for the caller to listen to all events, right? Because receiving a notification is async it could have happened that in-between a notification being triggered, subscriber task being waken up, another event (upload) happened.

Correct?